pkg/piecestore/rpc/server/ttl/ttl.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package ttl
+
+import (
+	"database/sql"
+	"fmt"
+	"log"
+	"time"
+
+	_ "github.com/mattn/go-sqlite3" // sqlite is weird and needs underscore
+
+	"storj.io/storj/pkg/piecestore"
+)
+
+// TTL -- ttl database
+type TTL struct {
+	DB *sql.DB
+}
+
+// NewTTL -- creates ttl database and struct
+func NewTTL(DBPath string) (*TTL, error) {
+	db, err := sql.Open("sqlite3", DBPath)
+	if err != nil {
+		return nil, err
+	}
+	_, err = db.Exec("CREATE TABLE IF NOT EXISTS `ttl` (`id` TEXT UNIQUE, `created` INT(10), `expires` INT(10));")
+	if err != nil {
+		return nil, err
+	}
+
+	return &TTL{db}, nil
+}
+
+// CheckEntries -- checks for and deletes expired TTL entries
+func CheckEntries(dir string, rows *sql.Rows) error {
+
+	for rows.Next() {
+		var expID string
+		var expires int64
+
+		err := rows.Scan(&expID, &expires)
+		if err != nil {
+			return err
+		}
+
+		// delete file on local machine
+		err = pstore.Delete(expID, dir)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("Deleted file: %s\n", expID)
+		if rows.Err() != nil {
+
+			return rows.Err()
+		}
+	}
+
+	return nil
+}
+
+// DBCleanup -- go routine to check ttl database for expired entries
+// pass in database and location of file for deletion
+func (ttl *TTL) DBCleanup(dir string) error {
+
+	tickChan := time.NewTicker(time.Second * 5).C
+	for {
+		select {
+		case <-tickChan:
+			rows, err := ttl.DB.Query(fmt.Sprintf("SELECT id, expires FROM ttl WHERE expires < %d", time.Now().Unix()))
+			if err != nil {
+				return err
+			}
+
+			err = CheckEntries(dir, rows)
+			if err != nil {
+				rows.Close()
+				return err
+			}
+			rows.Close()
+
+			_, err = ttl.DB.Exec(fmt.Sprintf("DELETE FROM ttl WHERE expires < %d", time.Now().Unix()))
```

Comment:
a comment i gave cam already - this is a new call to time.Now().Unix() from the call on line 72. if CheckEntries takes 5 minutes, this line is going to delete more than CheckEntries actually processed (if data expires in those 5 minutes). we need to save the timestamp at the top of the for loop and use it for the rest of the iteration

Reply:
👍